### ISSUE TITLE
Preact adding in environmental variables from the system is running on

### ISSUE THREAD

**Issue by scotlyt** [preactjs/preact-cli#1718]
**Preact recognizing system environmental variables so secrets can stay hidden in memory**

Hi all, 

I've combed over Preact documentation, looking at issues and I feel like I'm hitting a wall on trying to get Preact working in a production system. It doesn't appear that preact can recognize system environmental variables. It can recognize environmental variables if they are set by a .env or by loading them in with the preact.config.js file, but both of these ways are built on it being a file, not a system environmental variable. There doesn't really seem to be a clear cut way to hide environmental parts except through files which I'm trying to avoid.

Inside the docs on how to use environmental variable:
https://github.com/preactjs/preact-cli#using-environment-variables

> You can reference and use any environment variable in your application that has been prefixed with PREACT_APP_ automatically:
> 
>> src/index.js
> 
`console.log(process.env.PREACT_APP_MY_VARIABLE);`
> If your variable is not prefixed, you can still add it manually by using your preact.config.js (see [DefinePlugin](https://github.com/preactjs/preact-cli/wiki/Config-Recipes#use-environment-variables-in-your-application) config in the recipes wiki).
> 
> You can set and store variables using a .env file in the root of your project:
> 
>> .env
> 
`PREACT_APP_MY_VARIABLE="my-value"`
> You can also reference environment variables in your preact.config.js:
> 
```
 export default (config, env, helpers, options) => {
 	if (process.env.MY_VARIABLE) {
 		/** You can add a config here that will only used when your variable is truthy **/
 	}
```
> 
> };

**Additional context**

If we are trying to build on a system that defines everything inside a codebase, that means any privy information would have to exist as a form of a file where as if were to load them as a system environmental variable, we could load them in through a scripting process to pull them in from a secrets manager, put them in a tmp fs that only exists in memory and keeping the security of privy information much more secure since the secrets can live outside the codebase and the scripting process can load them in through memory instead of being compromised if an API key lives in a .env file.

I saw this on [stackexchange](https://stackoverflow.com/questions/70084981/how-to-hide-api-key-in-preact) that @rschristian answered and it leads me to believe that preact is not able to recognize system environmental variables.

I also saw more context that DefinePlugin is just acting as a search and replace through this [issue](https://github.com/preactjs/preact-cli/issues/1706#issuecomment-1178260623)

I'm not a preact expert and I haven't really coded with it, I'm really just trying to get preact working in a productional system that doesn't use a file to manage environment vars and would like to see this happen. If it does exist, feel free to close this and point me in the right direction.

**Comment by rschristian** [comment]
First of all, it's probably good to clarify that Preact != Preact-CLI. One is a UI lib, the other is a build tool / environment.

> It can recognize environmental variables if they are set by a `.env` or by loading them in with the `preact.config.js` file, but both of these ways are built on it being a file, not a system environmental variable. 

Not at all. `process.env.VARIABLE` is how you access env vars within Node. That has full access to system env vars.

> it leads me to believe that preact is not able to recognize system environmental variables.

Second example. `process.env.MY_VARIABLE` is a system environment variable.

Additionally, it might be good to specify that Preact-CLI creates static output. You cannot load the output of a web app onto a server and have it consume environment variables from the server. Everything is inlined at build time, though of course you could build on your production server if you really wanted to (would recommend against it though).

**Comment by scotlyt** [comment]
Gotcha, again this isn't really my area of expertise, I'm not a frontend engineer at all and only really dug into the weeds of frontend build process until recently.  Thank you.

**Comment by rschristian** [comment]
No worries!

If you can write up a more concrete example of what you want I'd be happy to write up examples/config or explain anything. This is a pretty esoteric part of modern web dev and varies quite a bit by tool.

**Comment by scotlyt** [comment]
Actually there was and was literally what I was digging into, we have an api.config.js that was defining where to send API calls from the preact app

```
const BASE_URL =
    process.env.PREACT_APP_ENVIRONMENT === "dev"
        ? process.env.PREACT_APP_API_URL_DEV
        : process.env.PREACT_APP_API_URL_PROD;

export { BASE_URL };
```

Looking at the package.json:

```json
    "scripts": {
        "build": "preact build --no-prerender",
        "serve": "sirv build --port 3016 --cors --single",
        "dev": "cross-env PREACT_APP_ENVIRONMENT=dev preact watch --port 3016",
```
Doing npm run dev works where the BASE_URL gives me what the env var specified.

However when doing npm run build and npm run serve, I got the BASE_URL as undefined. 

I was very confused and found that our preact.config.js was using `preact-cli-plugin-env-vars` which is an outdated plugin and lead me down rabbitholes of trying to get the preact.config.js loading with environment vars so I could create a docker image to be running our frontend components, which lead into figuring out docker-compose env vars, which lead to how do I get this running with nginx and onto an Amazon ECS ecosystem, so I could load in env vars with a secrets manager in the build process. So it's kind of has been a mess and I'm trying to get this straightened out for our frontend team since I'm going to be defining a lot of how their app will interact with the rest of our system

**Comment by rschristian** [comment]
> Actually there was and was literally what I was digging into, we have an api.config.js that was defining where to send API calls from the preact app
> 
> ```
> const BASE_URL =
>     process.env.PREACT_APP_ENVIRONMENT === "dev"
>         ? process.env.PREACT_APP_API_URL_DEV
>         : process.env.PREACT_APP_API_URL_PROD;
> 
> export { BASE_URL };
> ```
> 
> Looking at the package.json:
> 
> ```json
>     "scripts": {
>         "build": "preact build --no-prerender",
>         "serve": "sirv build --port 3016 --cors --single",
>         "dev": "cross-env PREACT_APP_ENVIRONMENT=dev preact watch --port 3016",
> ```
> 
> Doing npm run dev works where the BASE_URL gives me what the env var specified.
> 
> However when doing npm run build and npm run serve, I got the BASE_URL as undefined.

Sounds like `process.env.PREACT_APP_API_URL_PROD` isn't defined in your system (sorry if that's already clear to you, can't tell).

> I was very confused and found that our preact.config.js was using `preact-cli-plugin-env-vars` which is an outdated plugin

Still works, actually, but a couple months ago I integrated its functionality into Preact-CLI directly. Better for us to maintain and not require installation of a 3rd-party plugins for something so simple.

> and lead me down rabbitholes of trying to get the preact.config.js loading with environment vars so I could create a docker image to be running our frontend components, which lead into figuring out docker-compose env vars, which lead to how do I get this running with nginx and onto an Amazon ECS ecosystem, so I could load in env vars with a secrets manager in the build process. So it's kind of has been a mess and I'm trying to get this straightened out for our frontend team since I'm going to be defining a lot of how their app will interact with the rest of our system

Yeah, this can definitely be summed up as a rabbithole.

In short, you just need to provide your environment variables to your build environment (usually a CI/CD system). Once `preact build` completes, it's just static files from there. You can copy them directly into your production server, without need to load up the prod server with env vars.

**Comment by scotlyt** [comment]
>Sounds like process.env.PREACT_APP_API_URL_PROD isn't defined in your system (sorry if that's already clear to you, can't tell).

Ya I was trying to get it to pass from container environment vars, but there was another script in between the docker build process and the actual launch of the application, so the issue is most likely because of that middle script not passing things properly now, thats on my end no question.

> In short, you just need to provide your environment variables to your build environment (usually a CI/CD system). Once `preact build` completes, it's just static files from there. You can copy them directly into your production server, without need to load up the prod server with env vars.

I see, I've trying to ensure that our preact app is able to send calls to a backend api that would handle things like user authentication, so I was made to understand that the app would have an API key embedded in it, but it feels like that would be part of a web server configuration as opposed to the application itself.  

The goal was to not have any API secrets inside any .env file and to have the frontend working as a docker container, so the container would have all the environment vars inside it's compose area if built locally, but if was built in our infrastructure as code area, it would pull secrets in with a `run-app.sh`

It actually feels like preact build I'm doing wouldn't be inside a container but as static content inside a CDN for dev (and eventually prod) 

Gonna have to dig into this a lot more.

**Comment by rschristian** [comment]
> I've trying to ensure that our preact app is able to send calls to a backend api that would handle things like user authentication, so I was made to understand that the app would have an API key embedded in it, but it feels like that would be part of a web server configuration as opposed to the application itself.

Could also proxy requests? You can have your web app make requests to (say) `/api/foo`, where your server (be that nginx or whatever) would intercept the `/api` and direct traffic elsewhere. That's a pretty common pattern, and avoids having the URL embedded in the app. Depends on your needs.

Good luck though! Even simple devops does my head in, so I don't envy your position.

**Comment by scotlyt** [comment]
> Could also proxy requests? You can have your web app make requests to (say) `/api/foo`, where your server (be that nginx or whatever) would intercept the `/api` and direct traffic elsewhere. That's a pretty common pattern, and avoids having the URL embedded in the app. Depends on your needs.

I was presented a VM with a global nginx with the frontend and the backend is wrapped in docker that has a DB, cache, and other components. The backend also had an nginx piece which I believe now is the global nginx part looking at the config and has `location /api/` to itself on a different port &

```conf
             location /static/ {
                    root /var/www/html/;
                    autoindex on;
             }
```

So I've been untangling a lot of this, mainly trying to get them into separated areas, fixing the node_modules so they were built within a container (not dependent on the host system) and removing the shared filesystem inside the actual container where the db schema updates were dependent on dev containers. 

It sounds to me like a lot of what they did for the frontend interacting with the backend actually looks reasonable except for a part where I saw cleartext for the password field inside the console for preact and I started really digging into more on how the frontend was being built.

**Comment by scotlyt** [comment]
@rschristian wanted to follow up to all this

there wasn't any documentation that the assets being built in the one repo were all being copied into folder in another repo that would serve as all the static assets. The issue was that there was already some assets inside there and I was lead to believe that it was two different web apps talking to each other at the start.

Chalk it up to miscommunications and different timezones on working with people more than anything else.
